documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+respective materialized views. Because the schema of the media tables now
+matches what the ingestion server already expects, we can simply change the
+table name each place it is referenced:
+
+- [`_get_shared_cols`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L75)
+- [`refresh_api_table`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L310)
+- [`get_copy_data_query`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/queries.py#L206)
+
+## Dependencies
+
+### Tools & packages
+
+<!-- Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+We should not need any additional dependencies.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
+
+This does not depend on any existing projects.
+
+### Infrastructure
+
+<!-- Note any infrastructure this plan is dependent on. -->
+
+This project directly affects our Catalog database schema, Airflow DAGs, and
+ingestion server. All necessary changes are described in this document.
+
+Notably, because this plan would allow us to drop the materialized `image_view`
+and `audio_view`, we should net some DB storage savings in the long term.
+
+## Alternatives
+
+### Duplicate table
+
+Another approach was considered which, rather than updating records directly on
+the `image` table, would have instead created a duplicate table with updated
+scores and ultimately swapped tables. This has the benefit of avoiding
+concurrent writes to the `image` table, but the following disadvantages:
+
+- While the duplicate table is being built, writes to the `image` table from
+  provider DAGs would need to be propagated to it via a `Trigger`. Thus the same
+  deadlocking precautions are necessary, and the writes must still be batched,
+  offering no improved performance.
+- This approach requires much more space for the duplicate table (although
+  roughly equivalent to the current materialized view approach).
+- This approach is much less efficient because it unnecessarily re-inserts
+  **all** rows, including those that do not support popularity data at all.
+
+### Separate `image_popularity` view
+
+We considered keeping the `image` table as is, and an `image_popularity`
+materialized view equivalent to the current `image_view`. A new,
+non-materialized `image_view` would join `image` and `image_popularity` and be
+used for the data refresh. The advantage is that the `image_popularity` view
+could be refreshed independently of the data refresh; this approach would also
+be very quick and easy to implement. The disadvantages:
+
+- Would require an expensive `join` at the time of data refresh
```

Comment:
What is expensive about the join? Wouldn't the identifiers be indexed?

Reply:
I should probably remove the word 'expensive' since it does imply there's something _particularly_ expensive about this join, which I don't think is true. It is more that there is a preference to avoid any `join`s on the image table when possible.